sdk/core/core/src/az_curl_adapter.c

```diff
@@ -10,28 +10,248 @@
-  // create callback for visitor
-  az_pair_visitor const pair_visitor = az_headers_to_curl_callback(p_headers);
+/**
+ * @brief This is the function that curl will use to write response into a user provider span
+ * Function receives the size of the response and must return this same number, otherwise it is
+ * consider that function failed
+ *
+ * @param contents
+ * @param size
+ * @param nmemb
+ * @param userp
+ * @return int
+ */
+int write_to_span(void * contents, size_t size, size_t nmemb, void * userp) {
+  size_t const realsize = size * nmemb + 1;
+  az_span * const user_buffer = (az_span * const)userp;
+
+  // handle error when response won't feat user buffer
+  if (user_buffer->size < realsize) {
+    fprintf(stderr, "response size is greater than user buffer for writing response");
+    return AZ_ERROR_HTTP_FAILED_REQUEST;
+  }
+
+  // TODO: format buffer with AZ_RESPONSE_BUILDER
+  memcpy(&(user_buffer->begin[0]), contents, realsize);
+  // add 0 so response can be printed
+  user_buffer->begin[user_buffer->size] = 0;
+
+  // This callback needs to return the response size or curl will consider it as it failed
+  return (int)realsize - 1;
+}
+
+/**
+ * handles GET request
+ */
+az_result az_curl_send_request(
+    az_curl * const p_curl,
+    az_http_request_builder * const p_hrb,
+    az_span const * const response) {
+  // creates a slist for bulding curl headers
+  az_curl_headers_list headers = {
+    .p_list = NULL,
+  };
+  // build headers into a slist as curl is expecting
+  AZ_RETURN_IF_FAILED(az_build_headers(p_hrb, &headers));
+  // set all headers from slist
+  AZ_RETURN_IF_CURL_FAILED(curl_easy_setopt(p_curl->p_curl, CURLOPT_HTTPHEADER, headers.p_list));
 
-  az_pair_seq const request_headers_seq = p_request->headers;
-  AZ_RETURN_IF_FAILED(request_headers_seq.func(request_headers_seq.data, pair_visitor));
+  // set URL as 0-terminated str
+  size_t const extra_space_for_zero = (size_t)sizeof("\0");
+  size_t const url_final_size = p_hrb->url.size + extra_space_for_zero;
+  // allocate buffer to add \0
+  uint8_t * const p_writable_buffer = (uint8_t * const)malloc(url_final_size);
+  if (p_writable_buffer == NULL) {
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
+  // write url in buffer (will add \0 at the end)
+  char * buffer = (char *)p_writable_buffer;
+  az_span const writable_buffer
+      = (az_span const){ .begin = p_writable_buffer, .size = url_final_size };
+  az_result const result = az_write_url(writable_buffer, az_span_to_const_span(p_hrb->url));
+  CURLcode const set_headers_result = curl_easy_setopt(p_curl->p_curl, CURLOPT_URL, buffer);
+  // free used buffer before anything else
+  memset(p_writable_buffer, 0, url_final_size);
+  free(buffer);
+
+  // handle writing to buffer error
+  if (az_failed(result)) {
+    return result;
+  }
+  // handle setting curl url
+  if (set_headers_result != CURLE_OK) {
+    return AZ_ERROR_HTTP_FAILED_REQUEST;
+  }
```

Comment:
AZ_RETURN_IF_CURL_FAILED?

Reply:
got it! yes, I can use it like that. Thank you!  👍